src/Common/GuardedPoolAllocatorCrashHandler.cpp

```diff
@@ -0,0 +1,285 @@
+#include <Common/GuardedPoolAllocatorCommon.h>
+#include <Common/GuardedPoolAllocatorCrashHandler.h>
+#include <Common/StackTrace.h>
+#include <Common/logger_useful.h>
+
+#include <base/getThreadId.h>
+
+#include <cassert>
+#include <cinttypes>
+#include <cstring>
+
+namespace
+{
+struct ScopedEndOfReportDecorator
+{
+    explicit ScopedEndOfReportDecorator(Poco::Logger * log_) : log(log_) { LOG_FATAL(log, "*** GWP-ASan detected a memory error ***"); }
+    ~ScopedEndOfReportDecorator() { LOG_FATAL(log, "*** End GWP-ASan report ***\n"); }
+    Poco::Logger * log;
+};
+
+void printBacktrace(Poco::Logger * log, const StackTrace & trace)
+{
+    trace.toStringEveryLine([&](const std::string & s) { LOG_FATAL(log, fmt::runtime(s)); });
+}
+}
+
+namespace clickhouse_gwp_asan
+{
+
+using AllocationMetadata = clickhouse_gwp_asan::AllocationMetadata;
+using Error = clickhouse_gwp_asan::Error;
+
+CrashReporter::CrashReporter(
+    const clickhouse_gwp_asan::GuardedPoolAllocatorState * allocator_state_,
+    const clickhouse_gwp_asan::AllocationMetadata * allocator_metadata_)
+    : log(&Poco::Logger::get("GWP-Asan reporter")), allocator_state(allocator_state_), allocator_metadata(allocator_metadata_)
+{
+}
+
+bool CrashReporter::errorIsMine(uintptr_t error_ptr)
+{
+    assert(allocator_state && "allocator_state should not be nullptr.");
+    if (allocator_state->failure_type != Error::UNKNOWN && allocator_state->failure_address != 0)
+        return true;
+
+    return error_ptr < allocator_state->guarded_page_pool_end && allocator_state->guarded_page_pool <= error_ptr;
```

Comment:
Yes, but here `error_ptr` is `void *` and `pointerIsMine` has this signature: 
`bool pointerIsMine(const void * ptr) const;`

I guess we could call `state.pointerIsMine`, but we'd have to `reinterpret_cast<void *>(error_ptr)` beforehand, but I am not sure that it will be any better this way.

Reply:
You mean error_ptr is `uintptr_t` not `void *`.
But you explicitly cast to `uintptr_t` [here](https://github.com/ClickHouse/ClickHouse/pull/36826/files#diff-8e427f3323f6309a2cd8871847370fde2352316e0e3bfe3fab7d3ee214bae18dR287), what for? Maybe it is better to remove this cast and accept `void*` here too?

Reply:
reinterpret_cast is fine here btw.